sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScanBuilder.scala

```diff
@@ -68,6 +68,10 @@ case class ParquetScanBuilder(
   // All filters that can be converted to Parquet are pushed down.
   override def pushedFilters(): Array[Filter] = pushedParquetFilters
 
+  override def pruneColumns(requiredSchema: StructType): Unit = {
+    this.requiredSchema = requiredSchema
+  }
```

Comment:
for a datasource, how does it know passed in requiredSchema is for normal column pruning or nested column pruning?

I have this question because for a datasource that does not support nested column pruning, looks like when SQLConf.get.nestedSchemaPruningEnabled is true, a nested pruning required schema is still passed in. How does such datasource know to act?

Reply:
I think data sources should do their best to prune whatever they can.

Here is a quote from `SupportsPushDownRequiredColumns`:

> Implementation should try its best to prune the unnecessary columns or nested fields, but it's also OK to do the pruning partially, e.g., a data source may not be able to prune nested fields, and only prune top-level columns.

Reply:
Basically, the default implementation in `FileScanBuilder` is pruning only top-level attributes.